File: 2021/Day 02/main.js
// https://adventofcode.com/2021/day/2
'use strict';
const assert = require('assert');
const fs = require('fs');

const TASKS = [
	{ filename: "test1.txt", exp1: 150, exp2: 900 },
	{ filename: "input.txt" },
];

class Position {
	constructor() {
		this.x = 0;
		this.y = 0;
	}

	forward(v) {
		this.x += v;
	}

	down(v) {
		this.y += v;
	}

	up(v) {
		this.y -= v;
	}

	followDirections(directions) {
		directions.forEach(mov => {
			const [dir, v] = mov.split(' ');
			assert(v > 0);

			switch(parseDir(dir)) {
				case Direction.FORWARD: this.forward(parseInt(v)); break;
				case Direction.DOWN: this.down(parseInt(v)); break;
				case Direction.UP: this.up(parseInt(v)); break;
				default: throw new TypeError("Unreachable");
			}
		});
	}

	get result() {
		return this.x * this.y;
	}
}

class Submarine extends Position {
	constructor() {
		super();
		this.aim = 0;
	}

	forward(v) {
		super.forward(v);
		this.y += v * this.aim;
	}

	down(v) {
		this.aim += v;
	}

	up(v) {
		this.aim -= v;
	}
}

const Direction = {
	FORWARD: 0,
	DOWN: 1,
	UP: 2
}

function parseDir(d) {
	switch(d) {
		case "forward": return Direction.FORWARD;
		case "down": return Direction.DOWN;
		case "up": return Direction.UP;
		default: throw new TypeError("Unreachable");
	}
}

function part1(contents) {
	const submarine = new Position();
	submarine.followDirections(contents);

	return submarine.result;
}

function part2(contents) {
	const submarine = new Submarine();
	submarine.followDirections(contents);

	return submarine.result;
}

function solveFile({ filename, exp1, exp2, contents }) {
	console.log(`Solving file "${filename}:"`);
	if(contents === undefined) {
		contents = fs.readFileSync(filename, { encoding: 'utf8', flag: 'r' }).split('\n');
	}

	if(exp1 === undefined && exp2 === undefined) {	// Input file
		const res1 = part1(contents);
		const res2 = part2(contents);

		console.log(`Part 1: ${res1}`);	// 2150351
		console.log(`Part 2: ${res2}`);	// 1842742223
		return;
	}

	if(exp1 !== undefined) {
		const res1 = part1(contents);
		assert(res1 === exp1, `Expected ${exp1} but found ${res1}`);
		console.log(`Part1: ✅`);
	}
	if(exp2 !== undefined) {
		const res2 = part2(contents);
		assert(res2 === exp2, `Expected ${exp2} but found ${res2}`);
		console.log(`Part2: ✅`);
	}
}

for(let task of TASKS) {
	solveFile(task);
	console.log();
}

File: 2021/Day 02/main.hs
-- https://adventofcode.com/2021/day/2
import           System.IO   (readFile)
import           Text.Printf (printf)

type Point = (Int, Int)

initPoint :: Point
initPoint = (0, 0)

type DirPoint = (Int, Int, Int)

initDirPoint :: DirPoint
initDirPoint = (0, 0, 0)

data Dir = Forward Int | Down Int | Up Int

parseDir :: String -> Dir
parseDir s =
    case d of
        "forward" -> Forward v
        "down"    -> Down v
        "up"      -> Up v
        _         -> error $ "Unrecognized direction " ++ d
    where [d, val] = words s
          v = read val :: Int

applyDirection :: Point -> Dir -> Point
applyDirection (x, y) (Forward v) = (x + v, y    )
applyDirection (x, y) (Down v)    = (x,     y + v)
applyDirection (x, y) (Up v)      = (x,     y - v)

applyDirection2 :: DirPoint -> Dir -> DirPoint
applyDirection2 (x, y, aim) (Forward v) = (x + v, y + v * aim, aim    )
applyDirection2 (x, y, aim) (Down v)    = (x,     y,           aim + v)
applyDirection2 (x, y, aim) (Up v)      = (x,     y,           aim - v)

calcSolution :: Point -> Int
calcSolution (x, y) = x * y

calcSolution2 :: DirPoint -> Int
calcSolution2 (x, y, aim) = x * y

main :: IO ()
main = do
    directions <- map parseDir . lines <$> readFile "./input.txt"

    printf "Part 1: %d\n" $ calcSolution  $ foldl applyDirection  initPoint    directions
    printf "Part 2: %d\n" $ calcSolution2 $ foldl applyDirection2 initDirPoint directions
